### PR TITLE
Error handling

### LINKED ISSUE #27: handle errors
Once PR #23 is merged, we need to change the solution to show errors being handled. Particularly after [the confusion from last week](https://github.com/foundersandcoders/london-curriculum/blob/e8f0b7951dce505264e625fa4dc2395eea404bb1/stop-go-continue/fac-11/cohort-week-5.md#go) over the [issues from the `error-handling-workshop`](https://github.com/foundersandcoders/error-handling-workshop/issues/17)
+ https://github.com/pbywater/pg-workshop/blob/solution/src/router.js#L15
+ https://github.com/pbywater/pg-workshop/blob/solution/src/router.js#L24
+ https://github.com/pbywater/pg-workshop/blob/solution/src/router.js#L36
+ https://github.com/pbywater/pg-workshop/blob/solution/src/router.js#L41
+ https://github.com/pbywater/pg-workshop/blob/solution/src/router.js#L52

### PR DESCRIPTION
**Pull request by pbywater**
**Don't merge**

All requested changes have been made and pushed to PR #29 so I will close this now.

Fixes #27 
Adds error handling messages for users to see on front end.

### CODE REVIEW

File: solutions/src/queries/getData.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// Add code below to query your database
+const dbConnection = require('../database/db_connection.js');
+
+const getData = (cb) =>{
+    dbConnection.query(`SELECT * FROM users`
+    , (err, res) => {
+    if {
+      (err) cb(err);
```
Comment (eliasmalik) [inline solutions/src/queries/getData.js:8]: syntax error

File: solutions/src/queries/postData.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const dbConnection = require('../database/db_connection.js');
+
+const postData = (name, location, cb) => {
+  dbConnection.query('INSERT INTO users (name, location) VALUES ($1, $2)', [name, location], (err, res) => {
+    if (err) {
+      return cb(err);
+    }
+    else {
+    cb(null, res);
```
Comment (eliasmalik) [inline solutions/src/queries/postData.js:9]: indentation

File: solutions/src/router.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  const endpoint = request.url.split('/')[1];
+
+  if (endpoint === '') {
+    response.writeHead(200, { "Content-Type": "text/html" });
+    fs.readFile(__dirname + "/../public/index.html", function(error, file) {
+      if(error) {
+        response.writeHead(500, 'Content-Type:text/html');
+        response.end('<h1>Sorry, there was a problem loading the homepage</h1>');
+        console.log(error);
+        return;
```
Comment (eliasmalik) [inline solutions/src/router.js:18]: this `return` is unnecessary since you have an `else` block

File: solutions/src/router.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+const fs = require('fs');
+const pg = require('pg');
+const getData = require('./queries/getData.js');
+const postData = require('./queries/postData.js');
+
+const router = (request, response) => {
+  const endpoint = request.url.split('/')[1];
+
+  if (endpoint === '') {
+    response.writeHead(200, { "Content-Type": "text/html" });
```
Comment (eliasmalik) [inline solutions/src/router.js:12]: 1. Be consistent with how you're setting your headers. Here you use an object, in other places you set the string directly.
2. I would move this `writeHead` call into the callback. Seems odd to write it here and then overwrite it later.

File: solutions/src/router.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    });
+  } else if (endpoint === "users") {
+    // TASK 1: replace the 3 lines below below with your own function that gets data from your database
+    getData((err, res) => {
+        if (err) {
+          response.writeHead(500, 'Content-Type:text/html');
+          response.end('<h1>Sorry, there was a problem getting the users</h1>');
+          console.log(error);
+        };
+        let output = JSON.stringify(res);
```
Comment (eliasmalik) [inline solutions/src/router.js:31]: this needs to be in an `else` otherwise we'll try to respond twice

File: solutions/src/router.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    });
+    request.on('end', () => {
+      const name = queryString.parse(data).name;
+      const location = queryString.parse(data).location;
+
+      postData(name, location, (err, res) => {
+        if (err) {
+          response.writeHead(500, 'Content-Type:text/html');
+          response.end('<h1>Sorry, there was a problem adding that user</h1>');
+          console.log(err)};
```
Comment (eliasmalik) [inline solutions/src/router.js:50]: Looks like there's no response if data is successfully posted? the browser would hang? Also the closing brace of the if statement is not on a new line.

File: solutions/src/server.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+const http = require('http');
+const router = require('./router.js');
+
+const server = http.createServer(router);
+const port = process.env.PORT || 5000;
+
+const startServer = () => {
```
Comment (eliasmalik) [inline solutions/src/server.js:7]: This seems redundant. Why not just call `server.listen` straight away?

**Review by eliasmalik — CHANGES_REQUESTED** [review]
The indentations and code formatting are inconsistent in this file. Please run it through a linter.

**Comment by shiryz** [comment]
@pbywater can you please fix these either here or in #29 and then merge one into the other? 
it'll make reviewing easier if we don't have to look at the same code twice :heart: .

**Comment by pbywater** [comment]
@shiryz @eliascodes @jsms90 - all changes requested here have been made in PR #29 so will close this now

**Pull request closed without merging**